File: AudiobookManager/AudiobookManager.Domain/Audiobook.cs
namespace AudiobookManager.Domain;

public class Audiobook
{
    public long? Id { get; set; }
    public List<Person> Authors { get; set; }
    public List<Person> Narrators { get; set; }
    public string? BookName { get; set; }
    public string? Subtitle { get; set; }
    public string? Series { get; set; }
    public string? SeriesPart { get; set; }
    public int? Year { get; set; }
    public List<string> Genres { get; set; }
    public string? Description { get; set; }
    public string? Copyright { get; set; }
    public string? Publisher { get; set; }
    public string? Language { get; set; }
    public string? Rating { get; set; }
    public string? Asin { get; set; }
    public string? Www { get; set; }
    public AudiobookImage? Cover { get; set; }
    public string? CoverFilePath { get; set; }

    public int? DurationInSeconds { get; set; }
    public bool? ReplaceExisting { get; set; }

    /// <summary>
    /// UTC timestamp of the last time this book's metadata was applied/checked against an online
    /// source. Set only by AudiobookService.MarkMetadataRefreshedAsync - never from request DTOs
    /// and never written by the normal insert/update pipeline (it stays null there).
    /// </summary>
    public DateTime? LastMetadataRefreshedAt { get; set; }

    public AudiobookFileInfo FileInfo { get; set; }

    public Audiobook(List<Person> authors, string? bookName, int? year, AudiobookFileInfo fileInfo)
    {
        Authors = authors;
        BookName = bookName;
        Year = year;
        FileInfo = fileInfo;

        Narrators = new List<Person>();
        Genres = new List<string>();
    }
}

File: client/src/types/Audiobook.ts
import type { components } from "@/lib/api-types";
import type { Require } from "@/lib/dto";
import type BookFileInfo from "./BookFileInfo";

// AudiobookManager.Domain/AudiobookImage.cs: both fields are non-nullable.
export type AudiobookImage = Require<
  components["schemas"]["AudiobookImage"],
  "base64Data" | "mimeType"
>;

// Not generated: this is the frontend's own editing-form model, not a wire DTO. The backend
// saves/loads via flat "/"-joined author/genre strings (see
// helpers/organizeAudiobookInput.ts's splitList/joinList and services/api.ts's toAudiobookDto),
// while this shape is the richer, array-based representation BookEditForm and friends work
// with — services/api.ts converts between the two. Generating it from api-types.ts would just
// reproduce the DTO and lose that transform.
export interface Audiobook {
  authors: AudiobookPerson[];
  narrators: AudiobookPerson[];
  bookName?: string;
  subtitle?: string;
  series?: string;
  seriesPart?: string;
  year?: number;
  genres: string[];
  description?: string;
  copyright?: string;
  publisher?: string;
  language?: string;
  rating?: string;
  asin?: string;
  www?: string;

  cover?: AudiobookImage;

  durationInSeconds?: number;

  replaceExisting?: boolean;

  // Transient save-signal, not book data: "fields on this save were applied from a metadata
  // search result". Set by BookEditForm when TagPreviewDialog fields are applied, cleared
  // after the save that carried it; toAudiobookDto forwards it, toPathPreviewDto ignores it.
  metadataAppliedFromSearch?: boolean;

  // Client-only, never sent to the server (toAudiobookDto ignores it): marks a save as "apply
  // the stored metadata-refresh snapshot" so BookDetail can dismiss that snapshot after the
  // save completes. Rides on the audiobook object (rather than component state) so it is
  // discarded if the save is cancelled at the target-collision dialog instead of leaking into
  // the next unrelated save.
  pendingRefreshApplied?: boolean;

  fileInfo?: BookFileInfo;
}

// Not generated (see Audiobook above) — also used to construct locally-built author/narrator
// entries (e.g. `{ name }`) before they're flattened into the wire's joined-string form, so it
// deliberately omits the backend Person schema's `id` field, which the frontend never needs.
export interface AudiobookPerson {
  name: string;
  role?: string;
}

export type { Audiobook as default };
